engine/particles/private/particle_module.cpp

```diff
@@ -1,17 +1,177 @@
 #include "particle_module.hpp"
+
+#include "components/rigidbody_component.hpp"
+#include "components/transform_component.hpp"
+#include "components/transform_helpers.hpp"
+#include "components/world_matrix_component.hpp"
 #include "ecs_module.hpp"
+#include "emitter_component.hpp"
 #include "engine.hpp"
-#include "particle_interface.hpp"
+#include "gpu_resources.hpp"
+#include "graphics_context.hpp"
+#include "graphics_resources.hpp"
+#include "physics_module.hpp"
 #include "renderer.hpp"
-#include "renderer_module.hpp"
-
-#include <memory>
+#include "resource_management/image_resource_manager.hpp"
+#include "time_module.hpp"
+#include <renderer_module.hpp>
 
 ModuleTickOrder ParticleModule::Init(Engine& engine)
 {
-    auto& ecs = engine.GetModule<ECSModule>();
-    auto renderer = engine.GetModule<RendererModule>().GetRenderer();
-    _particleInterface = std::make_unique<ParticleInterface>(renderer->GetContext(), ecs);
+    _physics = &engine.GetModule<PhysicsModule>();
+    _context = engine.GetModule<RendererModule>().GetRenderer()->GetContext();
+    _ecs = &engine.GetModule<ECSModule>();
+
+    return ModuleTickOrder::ePostRender;
+}
+
+void ParticleModule::Tick(MAYBE_UNUSED Engine& engine)
+{
+    const auto emitterView = _ecs->GetRegistry().view<EmitterComponent>();
+    for (const auto entity : emitterView)
+    {
+        if (_ecs->GetRegistry().all_of<RigidbodyComponent>(entity))
+        {
+            const auto& rb = _ecs->GetRegistry().get<RigidbodyComponent>(entity);
+            if (_physics->bodyInterface->GetMotionType(rb.bodyID) != JPH::EMotionType::Static)
+            {
+                _ecs->GetRegistry().emplace_or_replace<ActiveEmitterTag>(entity);
+            }
+        }
+    }
+
+    const auto activeView = _ecs->GetRegistry().view<EmitterComponent, ActiveEmitterTag>();
+    for (const auto entity : activeView)
+    {
+        auto& emitter = _ecs->GetRegistry().get<EmitterComponent>(entity);
+
+        // first remove active tags from inactive emitters and continue
+        if (emitter.emitOnce)
+        {
+            _ecs->GetRegistry().remove<EmitterComponent>(entity);
+            _ecs->GetRegistry().remove<ActiveEmitterTag>(entity);
+            continue;
+        }
+        if (_ecs->GetRegistry().all_of<RigidbodyComponent>(entity))
+        {
+            const auto& rb = _ecs->GetRegistry().get<RigidbodyComponent>(entity);
+
+            // TODO: this is temporary to avoid when an entity doesn't have a WorldMatrixComponent. Would there be a way to do this differently?
```

Comment:
Doesn't every entity in the hierarchy guaranteed have a world matrix component?

Reply:
for some reason it's not guaranteed. I'm guessing newly created entities have their world matrix added after the particle module's tick since when I try to grab the world matrix component directly it just crashes